src/main/java/com/github/javabdd/JFactory.java

```diff
@@ -2377,6 +2425,974 @@ int zdiff_rec(int l, int r) {
         return res;
     }
 
+    int bdd_relnext(int states, int relation, int vars) {
+        // Check validity of BDD nodes.
+        CHECKa(states);
+        CHECKa(relation);
+        CHECKa(vars);
+
+        // Initialize caches if needed.
+        if (applycache == null) {
+            applycache = BddCacheI_init(cachesize);
+        }
+        if (itecache == null) {
+            itecache = BddCacheI_init(cachesize);
+        }
+
+        // We may also apply the OR operation while computing 'relnext'.
+        applyop = bddop_or;
+
+        int result;
+        int numReorder = 1;
+
+        again:
+        for (;;) {
+            try {
+                INITREF();
+
+                if (numReorder == 0) {
+                    bdd_disable_reorder();
+                }
+                result = relnext_rec(states, relation, vars);
+
+                if (numReorder == 0) {
+                    bdd_enable_reorder();
+                }
+            } catch (ReorderException x) {
+                bdd_checkreorder();
+                numReorder--;
+                continue again;
+            }
+            break;
+        }
+
+        checkresize();
+        return result;
+    }
+
+    int relnext_rec(int states, int relation, int vars) {
+        if (VERIFY_ASSERTIONS) {
+            _assert(!ZDD);
+        }
+
+        if (cachestats.enabled) {
+            cachestats.opAccess++;
+        }
+
+        // Terminals cases.
```

Comment:
what if states is not zero/one, and relation is one, then result is one, right? probably won't occur in practice though?

Reply:
That's only true if all leftover variables to consider in `states` are allowed to change. That is, it's only true if for every old-state BDD variable `x` that still occurs in `states`, it holds that the corresponding new-state variable `x'` is in `vars`.

Reply:
Ah, right. Never mind then.